File: server/src/retrieval/qdrantClient.js
// Thin REST wrapper around Qdrant. Not wired to a running instance yet —
// current retrieval still goes through rag.js's in-JS cosine search.
// Once a Qdrant container is up, point QDRANT_URL at it and swap
// retrievalService.js's semanticSearch() to use this client.
const QDRANT_URL = process.env.QDRANT_URL || 'http://127.0.0.1:6333';

export const COLLECTIONS = {
  documents: { name: 'nova_documents', vectorSize: 384, distance: 'Cosine' },
  code: { name: 'nova_code', vectorSize: 384, distance: 'Cosine' },
  web: { name: 'nova_web', vectorSize: 384, distance: 'Cosine' },
  media: { name: 'nova_media', vectorSize: 384, distance: 'Cosine' }
};

async function request(path, options = {}) {
  const res = await fetch(`${QDRANT_URL}${path}`, {
    headers: { 'Content-Type': 'application/json' },
    ...options
  });
  if (!res.ok) {
    const body = await res.text().catch(() => '');
    throw new Error(`Qdrant ${options.method || 'GET'} ${path} failed: ${res.status} ${body}`);
  }
  return res.json();
}

export async function ensureCollection(collection) {
  return request(`/collections/${collection.name}`, {
    method: 'PUT',
    body: JSON.stringify({
      vectors: { size: collection.vectorSize, distance: collection.distance }
    })
  });
}

export async function ensureAllCollections() {
  return Promise.all(Object.values(COLLECTIONS).map(ensureCollection));
}

export async function upsertPoints(collectionName, points) {
  // points: [{ id, vector, payload }]
  return request(`/collections/${collectionName}/points?wait=true`, {
    method: 'PUT',
    body: JSON.stringify({ points })
  });
}

export async function search(collectionName, vector, { limit = 10, filter } = {}) {
  const data = await request(`/collections/${collectionName}/points/search`, {
    method: 'POST',
    body: JSON.stringify({ vector, limit, filter, with_payload: true })
  });
  return data.result; // [{ id, score, payload }]
}

File: server/app/retrieval/qdrant_client.py
import httpx

from ..core.config import QDRANT_URL

COLLECTIONS = {
    "documents": {"name": "nova_documents", "vectorSize": 384, "distance": "Cosine"},
    "code": {"name": "nova_code", "vectorSize": 384, "distance": "Cosine"},
    "web": {"name": "nova_web", "vectorSize": 384, "distance": "Cosine"},
    "media": {"name": "nova_media", "vectorSize": 384, "distance": "Cosine"},
}


async def _request(path: str, method: str = "GET", json: dict | None = None) -> dict:
    url = f"{QDRANT_URL}{path}"
    async with httpx.AsyncClient(timeout=30) as client:
        res = await client.request(method, url, json=json, headers={"Content-Type": "application/json"})
        if res.status_code >= 400:
            raise RuntimeError(f"Qdrant {method} {path} failed: {res.status_code} {res.text}")
        return res.json()


async def ensure_collection(collection: dict):
    # Qdrant's create-collection PUT is not idempotent — it 409s if the
    # collection already exists (e.g. from a prior run). Treat that as success
    # instead of tearing down/recreating (would drop existing vectors).
    url = f"{QDRANT_URL}/collections/{collection['name']}"
    async with httpx.AsyncClient(timeout=30) as client:
        res = await client.put(
            url,
            json={"vectors": {"size": collection["vectorSize"], "distance": collection["distance"]}},
            headers={"Content-Type": "application/json"},
        )
        if res.status_code >= 400 and "already exists" not in res.text:
            raise RuntimeError(f"Qdrant PUT /collections/{collection['name']} failed: {res.status_code} {res.text}")
        return res.json() if res.status_code < 400 else {"status": "already_exists"}


async def ensure_all_collections():
    for collection in COLLECTIONS.values():
        await ensure_collection(collection)


async def upsert_points(collection_name: str, points: list[dict]):
    return await _request(
        f"/collections/{collection_name}/points?wait=true",
        method="PUT",
        json={"points": points},
    )


async def search(collection_name: str, vector: list[float], limit: int = 10, filter: dict | None = None):
    data = await _request(
        f"/collections/{collection_name}/points/search",
        method="POST",
        json={"vector": vector, "limit": limit, "filter": filter, "with_payload": True},
    )
    return data["result"]
